File: tools/excel/src/main/java/top/bettercode/summer/tools/excel/ExcelCell.java
package top.bettercode.summer.tools.excel;

/**
 * @author Peter Wu
 */
public class ExcelCell<T> {

  /**
   * 默认日期格式
   */
  public static final String DEFAULT_DATE_FORMAT = "yyyy-m-dd";
  /**
   * 默认时间格式
   */
  public static final String DEFAULT_DATE_TIME_FORMAT = "yyyy-m-dd hh:mm";
  /**
   * 默认格式
   */
  public static final String DEFAULT_FORMAT = "@";

  private final int row;
  private final int column;
  private final boolean lastRow;
  private boolean fillColor;
  private final Object cellValue;
  private final T entity;
  private final ExcelField<T, ?> excelField;

  public ExcelCell(int row, int column, int firstRow, boolean lastRow,
      ExcelField<T, ?> excelField, T entity) {
    this(row, column, lastRow, row - firstRow + 1, (row - firstRow + 1) % 2 == 0, excelField,
        entity);
  }

  public ExcelCell(int row, int column, boolean lastRow, int index, boolean fillColor,
      ExcelField<T, ?> excelField, T entity) {
    this.row = row;
    this.column = column;
    this.lastRow = lastRow;
    this.fillColor = fillColor;
    this.excelField = excelField;
    this.entity = entity;

    if (this.excelField.isIndexColumn()) {
      this.cellValue = index;
    } else {
      this.cellValue = excelField.toCellValue(entity);
    }
  }

  public void setFillColor(int index) {
    this.fillColor = index % 2 == 0;
  }

  public boolean needSetValue() {
    return true;
  }

  public int getRow() {
    return row;
  }

  public int getColumn() {
    return column;
  }

  public boolean isLastRow() {
    return lastRow;
  }


  public boolean isFillColor() {
    return fillColor;
  }

  public Object getCellValue() {
    return cellValue;
  }

  public T getEntity() {
    return entity;
  }

  public ExcelField<T, ?> getExcelField() {
    return excelField;
  }
}

File: tools/excel/src/main/kotlin/top/bettercode/summer/tools/excel/ExcelCell.kt
package top.bettercode.summer.tools.excel

import org.dhatim.fastexcel.BorderSide
import org.dhatim.fastexcel.BorderStyle
import org.dhatim.fastexcel.ProtectionOption
import java.math.BigDecimal
import java.time.LocalDate
import java.time.LocalDateTime
import java.time.ZonedDateTime
import java.util.*

/**
 *
 * @author Peter Wu
 */
class ExcelCell(
        private val excel: IExcel,
        private val style: CellStyle,
        private val row: Int,
        private val column: Int) {

    fun headerStyle(): ExcelCell {
        this.style.headerStyle()
        return this
    }

    fun setStyle(): ExcelCell {
        this.excel.setStyle(this.row, this.column, style)
        return this
    }

    fun width(width: Double): ExcelCell {
        this.excel.width(this.column, width)
        return this
    }

    fun height(height: Double): ExcelCell {
        this.excel.height(this.row, height)
        return this
    }

    fun formula(expression: String?): ExcelCell {
        this.excel.formula(this.row, this.column, expression)
        return this
    }

    fun comment(commen: String?): ExcelCell {
        this.excel.comment(this.row, this.column, commen)
        return this
    }

    fun value(): ExcelCell {
        this.excel.value(this.row, this.column)
        return this
    }

    fun value(value: String?): ExcelCell {
        this.excel.value(this.row, this.column, value)
        return this
    }

    fun value(value: Number?): ExcelCell {
        this.excel.value(this.row, this.column, value)
        return this
    }

    fun value(value: Boolean?): ExcelCell {
        this.excel.value(this.row, this.column, value)
        return this
    }

    fun value(value: Date?): ExcelCell {
        this.excel.value(this.row, this.column, value)
        return this
    }

    fun value(value: LocalDateTime?): ExcelCell {
        this.excel.value(this.row, this.column, value)
        return this
    }

    fun value(value: LocalDate?): ExcelCell {
        this.excel.value(this.row, this.column, value)
        return this
    }

    fun value(value: ZonedDateTime?): ExcelCell {
        this.excel.value(this.row, this.column, value)
        return this
    }

    fun dataValidation(dataValidation: Array<out String>): ExcelCell {
        this.excel.dataValidation(this.row, this.column, dataValidation)
        return this
    }

    fun style(style: CellStyle): ExcelCell {
        this.style.style(style)
        return this
    }

    fun format(numberingFormat: String?): ExcelCell {
        style.format(numberingFormat)
        return this
    }

    fun fillColor(rgb: String?): ExcelCell {
        style.fillColor(rgb)
        return this
    }

    fun fontColor(rgb: String?): ExcelCell {
        style.fontColor(rgb)
        return this
    }

    fun fontName(name: String?): ExcelCell {
        style.fontName(name)
        return this
    }

    fun fontSize(size: BigDecimal?): ExcelCell {
        style.fontSize(size)
        return this
    }

    fun fontSize(size: Int): ExcelCell {
        style.fontSize(size)
        return this
    }

    fun bold(): ExcelCell {
        style.bold()
        return this
    }

    fun italic(): ExcelCell {
        style.italic()
        return this
    }

    fun underlined(): ExcelCell {
        style.underlined()
        return this
    }

    fun horizontalAlignment(alignment: String?): ExcelCell {
        style.horizontalAlignment(alignment)
        return this
    }

    fun verticalAlignment(alignment: String?): ExcelCell {
        style.verticalAlignment(alignment)
        return this
    }

    @JvmOverloads
    fun wrapText(wrapText: Boolean = true): ExcelCell {
        style.wrapText(wrapText)
        return this
    }

    fun borderStyle(borderStyle: BorderStyle?): ExcelCell {
        style.borderStyle(borderStyle)
        return this
    }

    fun borderStyle(borderStyle: String?): ExcelCell {
        style.borderStyle(borderStyle)
        return this
    }

    fun borderStyle(side: BorderSide, borderStyle: BorderStyle): ExcelCell {
        style.borderStyle(side, borderStyle)
        return this
    }

    fun borderStyle(side: BorderSide, borderStyle: String): ExcelCell {
        style.borderStyle(side, borderStyle)
        return this
    }

    fun borderColor(borderColor: String?): ExcelCell {
        style.borderColor(borderColor)
        return this
    }

    fun borderColor(side: BorderSide, borderColor: String): ExcelCell {
        style.borderColor(side, borderColor)
        return this
    }

    fun protectionOption(option: ProtectionOption, value: Boolean): ExcelCell {
        style.protectionOption(option, value)
        return this
    }

}